apps/dashr-svm/utils/helperFunctions.R

```diff
@@ -0,0 +1,331 @@
+library(caret)
+library(kernlab)
+library(data.table)
+library(plotly)
+library(ROCR)
+
+# split the data into train and test sets
+trainTestSplit <- function(data){
+  data$Y <- as.factor(data$Y)
+  training_idx <- createDataPartition(y = data$Y, p = 0.6)
+  train_data <- data[training_idx[[1]]]
+  test_data <- data[!training_idx[[1]]]
+  return(list(train = train_data, test = test_data))
+}
+
+# Fit the model
+runSVM <- function(train_data, C = 1, sigma = 0.75, degree = 3,
+                   kernel = c("rbf", "linear", "poly", "sigmoid"),
+                   shrinking = TRUE){
+  if (kernel == "rbf"){
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "rbfdot",
+      C = C,
+      kpar = list(sigma = sigma),
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  } else if (kernel == "linear"){
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "vanilladot",
+      C = C,
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  } else if (kernel == "poly"){
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "polydot",
+      C = C,
+      kpar = list(degree = degree, scale = sigma, offset = 0),
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  } else if (kernel == "sigmoid"){
+    # tanh kernel equivalent to sklearn's sigmoid kernel (with correct settings)
+    svm <- ksvm(
+      as.matrix(train_data[, c("X1", "X2")]),
+      as.matrix(train_data$Y),
+      type = "C-svc",
+      kernel = "tanh",
+      C = C,
+      kpar = list(scale = sigma, offset = 0),
+      prob.model = TRUE,
+      shrinking = shrinking
+    )
+  }
+  return(svm)
+}
+
+# Plot the datasets and predictions
+generate_main_plot <- function(data, model, threshold){
+  train_data <- data[["train"]]
+  test_data <- data[["test"]]
+  full_data <- rbindlist(data)
+  y_pred_train <- as.factor(
+    as.integer(
+      predict(
+        model,
+        as.matrix(train_data[, c("X1", "X2")]),
+        type = "decision"
+      ) > threshold
+    )
+  )
```

Comment:
Since pipe operator `%>%` has already been loaded, we can write the code as:
`{predict(
      model,
      as.matrix(train_data[, c("X1", "X2")]),
      type = "decision"
) > threshold} 
%>% as.integer() 
%>% as.factor() -> y_pred_train`

Reply:
True, changing that to your format, it's a bit nicer to read (I hope)